server/src/main/java/org/apache/druid/server/coordinator/duty/MarkDanglingTombstonesAsUnused.java

```diff
@@ -0,0 +1,147 @@
+      RowKey datasourceKey = RowKey.of(Dimension.DATASOURCE, datasource);
+      stats.add(Stats.Segments.DANGLING_TOMBSTONE, datasourceKey, unusedSegments.size());
+      int unusedCount = deleteHandler.markSegmentsAsUnused(unusedSegments);
+      log.info(
+          "Successfully marked [%d] dangling tombstones as unused for datasource[%s].",
+          unusedCount,
+          datasource
+      );
+    });
+
+    return params;
+  }
+
+  /**
+   * Computes the set of dangling tombstones using the datasources snapshot.
+   *
+   * <li> Determine the set of used and non-overshadowed segments from the used segments' timeline. </li>
+   * <li> For each such segment that is dangling, look at the set of used segments in the timeline to see if the intervals overlap.
+   * There can at most be two such dangling tombstones per datasource. </li>
+   * <li> If there is no overlap, add the dangling segment to the result set to be marked as unused. </li>
+   * </p>
+   *
+   * @param dataSourcesSnapshot the datasources snapshot for segments timeline
+   * @return the set of dangling tombstones grouped by datasource
+   */
+  private Map<String, Set<SegmentId>> determineDanglingTombstones(final DataSourcesSnapshot dataSourcesSnapshot)
+  {
+    final Map<String, Set<SegmentId>> datasourceToDanglingTombstones = new HashMap<>();
+
+    dataSourcesSnapshot.getDataSourcesMap().keySet().forEach((datasource) -> {
+      final SegmentTimeline usedSegmentsTimeline
+          = dataSourcesSnapshot.getUsedSegmentsTimelinesPerDataSource().get(datasource);
+
+
+      final Optional<Set<DataSegment>> usedNonOvershadowedSegments =
+          Optional.fromNullable(usedSegmentsTimeline)
+                  .transform(timeline -> timeline.findNonOvershadowedObjectsInInterval(
+                      Intervals.ETERNITY,
+                      Partitions.ONLY_COMPLETE
+                  ));
+
+      if (usedNonOvershadowedSegments.isPresent()) {
+        usedNonOvershadowedSegments.get().forEach(usedNonOvershadowedSegment -> {
+          if (isDanglingTombstone(usedNonOvershadowedSegment)) {
+            boolean overlaps = usedSegmentsTimeline.iterateAllObjects().stream()
+                                                  .anyMatch(
+                                                      usedSegment ->
+                                                          datasource.equals(usedSegment.getDataSource()) &&
+                                                          !usedNonOvershadowedSegment.getId().equals(usedSegment.getId()) &&
+                                                          isDanglingTombstone(usedNonOvershadowedSegment) &&
+                                                          usedNonOvershadowedSegment.getInterval()
+                                                                                    .overlaps(usedSegment.getInterval())
+                                                  );
+            if (!overlaps) {
```

Comment:
So a dangling tombstone is one that is an infinity tombstone and DOES NOT overlap with another infinity tombstone?